### ISSUE TITLE
[Bug]: Duplicate <footer> on AMO version-history page, with its white text mostly-invisible but superimposed with the actual page content

### ISSUE THREAD

**Issue by dholbert** [mozilla/addons#14989]
### What happened?

If you visit https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions with a small window (e.g. 950px by 950px) and scroll down a little bit, you can sort of see there's some white text from a footer that's mysteriously present in the left sidebar that seems to overshoot onto the main page content.

The overshooting text is mostly white-on-white so it's mostly not visible, but it overlaps the actual content and causes artifacts.

If you look at the page DOM in devtools, you can see that this is actually a redundant `<footer>` element which is a next-sibling of the `#react-view` element. (There's another `<footer>` that's correctly placed, which is a descendant of the `#react-view` element.)

### What did you expect to happen?

No such redundant footer overlapping content.

### Is there an existing issue for this?

- [X] I have searched the existing issues

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/AMOENG-1014)

**Comment by dholbert** [comment]
Screenshots showing this out-of-place footer, clearly visible in the part overlapping the left sidebar, and somewhat-visible when overlapping content in the main area:
![image](https://github.com/user-attachments/assets/b45caa74-194a-49ea-baee-f5257bf39b53)
![image](https://github.com/user-attachments/assets/4db80282-abfd-4b9b-8c39-53f8c9589e34)
(note here^, the "Creative Commons" text is part of the text in the footer -- you can't see its surrounding text because it's white-text-on-white-background)
![image](https://github.com/user-attachments/assets/212a6604-f19f-420a-8c45-e62d0932e730)
(note here^, I've manually changed the footer text to have a red color instead of a white color so that you can see it all quite clearly)

**Comment by dholbert** [comment]
This doesn't reproduce on all version-history pages.  E.g. this one is fine:
https://addons.mozilla.org/en-US/firefox/addon/languagetool/versions/
...but this one is reliably broken (for me at least):
https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions/

**Comment by diox** [comment]
The problem is also slightly different when JS is off. It looks like something in the same vein of https://github.com/mozilla/addons/issues/14200, where the HTML is bad (unclosed element, element inside another element it's not supposed to be in, etc) and that React tries to reconcile its DOM poorly.

The fact that it only happens for this specific add-on is weird... maybe has to something to do with the content of their release notes.

**Comment by diox** [comment]
Bitwarden version `2024.4.2` has the following release notes:
```
<li>Extension requires additional browser permissions to better manage content script injection.</li>
<li>Small bug fixes and improvements</li>
```

We do allow developers to use `<li>` in there, but they should be enclosed in `<ul>`... On that page, each version is already in a `<li>` of its own, and I'm guessing that's what's causing the issue here.

**Comment by diox** [comment]
Can be reproduced at https://addons-dev.allizom.org/en-US/firefox/addon/ratopuse-weasp-vulturkal/versions/

You can see the content jump around if you reload that page, and the footer content being duplicated and displayed on top of the normal footer (or, depending on your window size, the main body content)

**Comment by diox** [comment]
Rough potential fix:

```javascript
  _purify.addHook(
    'uponSanitizeElement',
    function (node) {
      if (node.tagName === 'LI' && node.parentNode.tagName !== 'UL') {
          const oldParent = node.parentNode;
          const newParent = node.ownerDocument.createElement('ul');
          newParent.appendChild(node);
          oldParent.appendChild(newParent);
      }
    }
  );
````

But we need to do some refactoring to only do that in `sanitizeUserHTML` instead of the more global `sanitizeHTML`. The former doesn't have access to `DOMPurify` directly, hence the need for refactoring.

**Comment by diox** [comment]
`bleach` ensures tags are closed properly, so at least we don't have to worry about unbalanced tags.

We allow:
- `a`
- `abbr`
- `acronym`
- `b`
- `blockquote`
- `code`
- `em`
- `i`
- `li`
- `ol`
- `strong`
- `ul`

I've done various tests and only the `<li>` outside of a `<ul>` or `<ol>` is problematic. Technically, `blockquote`, `ul` or `ol` are not allowed inside the inline ones but in practice doesn't cause problems. We also accept any content inside `<ul>` and `<ol>` but again, that doesn't cause any issues.

So the only thing that really cause issues at this point are `<li>` without a `<ul>` or `<ol>` parent, which my approach would solve.

**Comment by diox** [comment]
We never prioritized that issue so it dropped from our radar but this is worth fixing.

**Comment by diox** [comment]
> But we need to do some refactoring to only do that in `sanitizeUserHTML` instead of the more global `sanitizeHTML`. The former doesn't have access to `DOMPurify` directly, hence the need for refactoring.

Actually... hooks are global, so it's a lot simpler to just apply this once and for all in `amo/purify.js`, where we create the `DOMPurify` instance that will get used everytime we call `sanitizeHTML`. Otherwise we'd have to have multiple instances around, I don't think it's worth it.

**Comment by diox** [comment]
Before, on [dev](https://addons-dev.allizom.org/en-US/firefox/addon/ratopuse-weasp-vulturkal/versions/):

![Image](https://github.com/user-attachments/assets/8c530ff1-493c-48e6-aa8c-f20a6a0b50f0)

After, once my patch lands, on dev:

![Image](https://github.com/user-attachments/assets/a7fd6d05-d355-4b9b-b8fe-74896ae3a85d)

**Comment by ioanarusiczki** [comment]
I remember I could reproduce this some while ago before replacing HTML with Markdown and if I remember it well it was only when I was directly landing on see all versions page and contained `<li>` in version notes.  So yes, I no longer see it on [-dev](https://addons-dev.allizom.org/en-US/firefox/addon/ratopuse-weasp-vulturkal/versions/).

Similarly with[ markdown on -stage](https://addons.allizom.org/en-US/firefox/addon/something-with-version-notes/versions/) this is not reproducible.

**Comment by diox** [comment]
Markdown makes it harder to reproduce, I fixed existing HTML content, as can be seen with https://addons-dev.allizom.org/en-US/firefox/addon/ratopuse-weasp-vulturkal/versions/

**Comment by ioanarusiczki** [comment]
Yes, thanks for the link , I've checked on that before making my comment.
